map.js

```diff
@@ -195,143 +183,97 @@ function drawMapOverlay(){
-    let mapPoint = new Point(point.x + mapOffset.x, point.y + mapOffset.y);
-    return Object.getPrototypeOf(Object.getPrototypeOf(this)).contains.call(this, mapPoint);
-}
-
-function drawIcon2(ctx){
-    //draws the name for the map icon if hovered.
-    if(this.cell.hive.classname != "nirnroot"){
-        if(this.contains(lastMouseLoc)){
-            //create rect that contains text and the icon
-            ctx.beginPath();
-            ctx.fillStyle = "#E5D9B9";
-            ctx.rect(this.minX-mapOffset.x, this.minY-mapOffset.y, (this.cell.name.length * 10) + this.width(), this.height());
-            ctx.fill();
-
-            ctx.beginPath();
-            ctx.fillStyle = "black";
-            ctx.textBaseline = "middle";
-            ctx.textAlign = "left";
-            ctx.font = "16px Monospace";
-            ctx.fillText(this.cell.name, this.maxX-mapOffset.x, this.minY + this.height() / 2 - mapOffset.y);
-            ctx.fill();
-        }
-    }
-    ctx.drawImage(this.icon, this.minX-mapOffset.x, this.minY-mapOffset.y, this.width(), this.height());
-    //if(discoveredArr.includes(locObj.formid)){
-    //    ctx.drawImage(icons.Check, this.minX, this.minY, this.width(), this.height());
-    //}
-}
-
 /*********************************
  * TOPBAR FUNCTIONS
  *  this is the "topbar" on the map canvas.
  *********************************/
 
 function initTopbar(){
-    function MapButton(x,y,width,height,text){
+    function MapButton(ordinal,y,height,text){
         MapObject.call(this);
-        this.minX = x;
+        this.ordinal = ordinal;
+        //this.minX and this.maxX calculated by recalculateBoundingBox
         this.minY = y;
-        this.maxX = x+width;
         this.maxY = y+height;
         this.name = text;
-        this.draw = function(ctx){
-            let width = this.width();
-            let height = this.height();
-            ctx.beginPath();
-            if(currentOverlay == this.name){
-                ctx.fillStyle = "#ccc";
-            }
-            else{
-                ctx.fillStyle = "#E5D9B9";
-            }
-            ctx.fillRect(this.minX, this.minY, width, height);
-
-            //and now text
-            ctx.beginPath();
-            ctx.fillStyle = "#000000";
-            ctx.textAlign = "center";
-            ctx.textBaseline = "alphabetic";
-            ctx.font = "16px Arial"
-            ctx.fillText(this.name, this.minX + (width/2), this.minY + 16);
-        }
+        this.recalculateBoundingBox();
+        
     }
     MapButton.prototype = Object.create(MapObject.prototype);
+    MapButton.prototype.draw = function(ctx){
+        let width = this.width();
+        let height = this.height();
+        ctx.beginPath();
+        if(currentOverlay == this.name){
+            ctx.fillStyle = "#ccc";
+        }
+        else{
+            ctx.fillStyle = "#E5D9B9";
+        }
+        ctx.fillRect(this.minX, this.minY, width, height);
```

Comment:
This is basically a copy of your code, with a little bit of commonality factored out.